File: L404-SumOfLeftLeaves.java
// https://leetcode.com/problems/sum-of-left-leaves/


class Solution {
    public static int search(TreeNode node, int sum, TreeNode parent) {
        if (node == null) {
            return sum;
        }
        if (parent != null && parent.left == node) {
            if (node.left == null && node.right == null) {
                sum += node.val;
            }
        }
        parent = node;
        sum = search(node.left, sum, parent);
        sum = search(node.right, sum, parent);
        return sum;
    }
    public int sumOfLeftLeaves(TreeNode root) {
        return search(root, 0, null);
    }
}

File: L404-sum_of_left_leaves.py
# https://leetcode.com/problems/sum-of-left-leaves/
class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        def search(root, add, parent):
            if (not root):
                return add
            if (parent and parent.left == root):
                if (not root.left and not root.right):
                    add += root.val
            parent = root
            add = search(root.left, add, parent)
            add = search(root.right, add, parent)
            return add
        return search(root, 0, None)
